examples/grpc-kotlin/src/main/kotlin/example/armeria/grpc/kotlin/HelloServiceImpl.kt

```diff
@@ -3,52 +3,56 @@ package example.armeria.grpc.kotlin
 import com.linecorp.armeria.server.ServiceRequestContext
 import example.armeria.grpc.kotlin.Hello.HelloReply
 import example.armeria.grpc.kotlin.Hello.HelloRequest
-import io.grpc.stub.StreamObserver
-import java.time.Duration
-import java.util.concurrent.TimeUnit
-import reactor.core.publisher.Flux
-import reactor.core.scheduler.Schedulers
-
-class HelloServiceImpl : HelloServiceGrpc.HelloServiceImplBase() {
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.withContext
+
+
+/**
+ * Note that if you want to run [HelloServiceImpl] in an event loop and
+ * get a current [ServiceRequestContext],
+ * you should initialize [HelloServiceImpl] with [Dispatchers.Unconfined].
+ * See https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-unconfined.html
+ * Unless [Dispatchers.Default] executor will handle this service.
+ */
+class HelloServiceImpl : HelloServiceGrpcKt.HelloServiceCoroutineImplBase(Dispatchers.Unconfined) {
 
     /**
      * Sends a [HelloReply] immediately when receiving a request.
      */
-    override fun hello(request: HelloRequest, responseObserver: StreamObserver<HelloReply>) {
-        responseObserver.onNext(buildReply(toMessage(request.name)))
-        responseObserver.onCompleted()
+    override suspend fun hello(request: HelloRequest): HelloReply {
+        // Make sure that current thread is request context aware
+        ServiceRequestContext.current()
+        return buildReply(toMessage(request.name))
     }
 
-    override fun lazyHello(request: HelloRequest, responseObserver: StreamObserver<HelloReply>) {
-        // You can use the event loop for scheduling a task.
-        ServiceRequestContext.current().contextAwareEventLoop().schedule({
-            responseObserver.onNext(buildReply(toMessage(request.name)))
-            responseObserver.onCompleted()
-        }, 3, TimeUnit.SECONDS)
+    override suspend fun lazyHello(request: HelloRequest): HelloReply {
+        delay(3000L)
```

Comment:
Check context here too, I think similar to the flow below it's not correct - and it'd be quite annoying to have to propagate it here too despite being a coroutine

Reply:
Sure!